controllers/upgrade_controller.go

```diff
@@ -0,0 +1,270 @@
+	storageClientOperatorVersionPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+
+			oldObj, _ := e.ObjectOld.(*v1alpha1.StorageClient)
+			newObj, _ := e.ObjectNew.(*v1alpha1.StorageClient)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if newObj.Status.Operator.CurrentVersion != newObj.Status.Operator.DesiredVersion {
+				return true
+			}
+
+			return false
+		},
+	}
+	ocsClientOperatorSubscriptionPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+
+			oldObj, _ := e.ObjectOld.(*opv1a1.Subscription)
+			newObj, _ := e.ObjectNew.(*opv1a1.Subscription)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if oldObj.Spec.Package != ocsClientOperatorSubscriptionPackageName {
+				return false
+			}
+
+			return oldObj.Spec.Channel != newObj.Spec.Channel
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		Watches(&v1alpha1.StorageClient{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(storageClientOperatorVersionPredicate)).
+		Watches(&opv1a1.Subscription{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(ocsClientOperatorSubscriptionPredicate)).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=operators.coreos.com,resources=subscriptions,verbs=get;list;patch
+
+func (r *UpgradeReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+	r.log.Info("starting reconcile")
+
+	conditionStatus := metav1.ConditionTrue
+	conditionMessage := "No component reported errors."
+	result, err := r.reconcilePhases()
+	if err != nil {
+		// TODO (lgangava): we are being very restrictive as of now, not upgrading if we hit any error
+		// differentiate between reconciler errors and not upgradeable errors
+		conditionStatus = metav1.ConditionFalse
+		conditionMessage = fmt.Sprintf("Operator could not be upgraded due to: %v", err)
+		r.log.Error(err, "an error occurred during reconcilePhases")
+	}
+	err = errors.Join(err, r.setOperatorCondition(conditionStatus, conditionMessage))
+
+	r.log.Info("reconciling completed")
+	return result, err
+}
+
+func (r *UpgradeReconciler) reconcilePhases() (ctrl.Result, error) {
+	if err := r.reconcileSubscriptionChannel(); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *UpgradeReconciler) reconcileSubscriptionChannel() error {
+	storageClientList := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClientList); err != nil {
+		return fmt.Errorf("failed to list storageclients %v", err)
+	}
+
+	if len(storageClientList.Items) == 0 {
+		r.log.Info("no storageclients exist")
+		return nil
+	}
+
+	desiredVersions := make([]string, 0, len(storageClientList.Items))
```

Comment:
Better to declare as a nil slice with `var`

Reply:
nope, if we know the capacity why to declare without it?